File: fushi/test/lookup/browser_extension_page_scroll_bug732_test.js
// BUG-732 behavior test: the shared lookup popup wheel handler must NOT hijack a
// normal web page's scrolling when Hibiki's browser extension is installed.
//
// Root cause it locks: popup.js is injected as a content script on <all_urls>.
// Its document 'wheel' listener re-implements scrolling for the *dictionary
// popup* (scaled by POPUP_WHEEL_PIXEL_FACTOR = 0.24, with preventDefault). When
// the wheel is over the host page (not the popup shadow) the handler used to fall
// through to `window.scrollBy` — commandeering every normal page's scroll at 24%
// speed. The fix: in the extension content-script context (chrome.runtime.id
// present) a wheel that is NOT over the popup shadow must fall through to native
// scrolling (no preventDefault). The in-app popup (flutter_inappwebview, no
// chrome) and wheels over the popup itself keep the custom scaled scroll.
//
// BUG-1078 extends this: even a cheap non-passive document wheel listener makes
// the browser abandon the compositor fast-scroll path on EVERY page the content
// script touches. So in the extension context popup.js must NOT attach to
// document at all — it exposes the handler as window.__fushiPopupWheelListener
// and content.js lazily mounts it on the popup shadow host for the popup's
// lifetime. The in-app popup (no chrome.runtime) keeps the resident document
// listener ({passive:false}) — there the document IS the popup.
//
// This EXECUTES the real popup.js wheel handler against a minimal fake DOM and
// asserts preventDefault / scrollBy per surface. Reverting the guard turns it red.
//
// Run: node fushi/test/lookup/browser_extension_page_scroll_bug732_test.js
// (also driven from browser_extension_page_scroll_bug732_test.dart inside
//  `flutter test`).

const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const popupPath = path.resolve(__dirname, '../../assets/popup/popup.js');
const source = fs.readFileSync(popupPath, 'utf8');

// Build a sandbox, load popup.js, and return the captured wheel handler plus
// recorders for the behaviours we assert on.
function loadWheelHandler(opts) {
  const calls = {
    prevented: false,
    windowScrollBy: [],
    hostScrollBy: [],
  };

  const documentWheelRegs = [];

  const hostEl = {
    style: {},
    scrollBy(arg) { calls.hostScrollBy.push(arg); },
  };

  const documentObj = {
    documentElement: { style: {} },
    body: { tagName: 'BODY' },
    addEventListener(type, handler, options) {
      if (type === 'wheel') documentWheelRegs.push({ handler, options });
    },
  };

  const windowObj = {
    innerHeight: 800,
    addEventListener() {},
    scrollBy(arg) { calls.windowScrollBy.push(arg); },
    getSelection() { return { toString() { return ''; } }; },
    getComputedStyle() { return {}; },
  };
  // __fushiRoot.host is the shadow host; only present when a popup is open.
  windowObj.__fushiRoot = opts.popupOpen ? { host: hostEl } : undefined;
  if (opts.flutter) {
    windowObj.flutter_inappwebview = { callHandler() { return Promise.resolve(false); } };
  }

  const sandbox = {
    Node: { TEXT_NODE: 3, ELEMENT_NODE: 1 },
    Date, Math, URL, JSON, RegExp, Set, Map, Object, Array, Number, Boolean,
    String, console,
    performance: { now() { return 1000; } },
    setTimeout, clearTimeout,
    DOMParser: class { parseFromString() { return { body: {}, querySelectorAll() { return []; } }; } },
    document: documentObj,
    window: windowObj,
    getComputedStyle() { return {}; },
  };
  if (opts.extension) {
    sandbox.chrome = { runtime: { id: 'hibiki-extension-id' } };
  }
  sandbox.globalThis = sandbox;

  vm.createContext(sandbox);
  vm.runInContext(source, sandbox, { filename: 'popup.js' });

  let wheelHandler = null;
  if (opts.extension) {
    // BUG-1078: extension context must leave document COMPLETELY free of wheel
    // listeners (a resident non-passive listener kills compositor fast scroll on
    // every page). The handler is exposed for content.js to mount on the shadow
    // host for the popup's lifetime.
    assert.strictEqual(documentWheelRegs.length, 0,
      'extension context must NOT attach any document wheel listener (BUG-1078)');
    wheelHandler = windowObj.__fushiPopupWheelListener;
    assert.ok(typeof wheelHandler === 'function',
      'extension context must expose window.__fushiPopupWheelListener for ' +
      'content.js to lazily mount on the popup shadow host');
  } else {
    // In-app popup WebView: the document IS the popup — the resident document
    // listener must stay, and stay non-passive (it preventDefault()s).
    assert.strictEqual(documentWheelRegs.length, 1,
      'in-app popup must keep exactly one resident document wheel listener');
    assert.ok(documentWheelRegs[0].options &&
      documentWheelRegs[0].options.passive === false,
      'in-app document wheel listener must be {passive:false} (it preventDefaults)');
    wheelHandler = documentWheelRegs[0].handler;
  }

  return { wheelHandler, calls, hostEl };
}

// A vertical mouse-notch wheel event over `target`; composedPath is [target,
// ...ancestors]. deltaMode 0 (pixels) so deltaPx === deltaY.
function makeWheelEvent(target, ancestors, calls) {
  const pathArr = [target].concat(ancestors || []);
  return {
    ctrlKey: false,
    deltaY: 100,
    deltaX: 0,
    deltaMode: 0,
    target,
    composedPath() { return pathArr; },
    preventDefault() { calls.prevented = true; },
  };
}

// Case A — the bug: extension installed, wheel over the HOST PAGE (no popup /
// wheel not over the popup). Native scrolling must win: no preventDefault, no
// custom window.scrollBy.
(function extensionHostPageFallsThroughToNative() {
  const { wheelHandler, calls } = loadWheelHandler({ extension: true, popupOpen: false });
  const target = { nodeType: 1, parentElement: null };
  wheelHandler(makeWheelEvent(target, [], calls));

  assert.strictEqual(calls.prevented, false,
    'extension host-page wheel must NOT call preventDefault (native scroll)');
  assert.strictEqual(calls.windowScrollBy.length, 0,
    'extension host-page wheel must NOT run the scaled window.scrollBy');
})();

// Case A2 — the bug persists even while a popup is open, as long as the wheel is
// over the host page (composedPath does not cross the shadow host).
(function extensionHostPageWithPopupOpenStillNative() {
  const { wheelHandler, calls } = loadWheelHandler({ extension: true, popupOpen: true });
  const target = { nodeType: 1, parentElement: null };
  wheelHandler(makeWheelEvent(target, [], calls)); // path never includes hostEl

  assert.strictEqual(calls.prevented, false,
    'wheel over host page (popup open) must still fall through to native');
  assert.strictEqual(calls.windowScrollBy.length, 0,
    'wheel over host page (popup open) must not scroll the window custom');
})();

// Case B — extension, wheel OVER the popup shadow: the custom scaled scroll must
// still drive the shadow host (the popup's real scroll container). Unchanged.
(function extensionWheelOverPopupScrollsHost() {
  const { wheelHandler, calls, hostEl } = loadWheelHandler({ extension: true, popupOpen: true });
  const inner = { nodeType: 1, parentElement: null };
  // composedPath crosses the shadow host -> __fushiWheelScroller returns it.
  wheelHandler(makeWheelEvent(inner, [hostEl], calls));

  assert.strictEqual(calls.prevented, true,
    'wheel over the popup must preventDefault (custom scaled scroll)');
  assert.strictEqual(calls.hostScrollBy.length, 1,
    'wheel over the popup must scroll the shadow host');
  assert.strictEqual(calls.windowScrollBy.length, 0,
    'wheel over the popup must not scroll the window');
  const step = calls.hostScrollBy[0].top;
  assert.ok(step > 0 && step <= 120,
    'popup scroll step must be the scaled/clamped delta, got ' + step);
})();

// Case C — the in-app popup document (flutter_inappwebview, no chrome): the whole
// document IS the popup, so window.scrollBy custom scroll must keep working. This
// is the regression guard: the fix must not disable in-app popup scrolling.
(function inAppPopupKeepsWindowScroll() {
  const { wheelHandler, calls } = loadWheelHandler({ extension: false, flutter: true, popupOpen: false });
  const target = { nodeType: 1, parentElement: null };
  wheelHandler(makeWheelEvent(target, [], calls));

  assert.strictEqual(calls.prevented, true,
    'in-app popup wheel must preventDefault (custom document scroll)');
  assert.strictEqual(calls.windowScrollBy.length, 1,
    'in-app popup wheel must drive window.scrollBy');
  const step = calls.windowScrollBy[0].top;
  assert.ok(step > 0 && step <= 120,
    'in-app popup scroll step must be the scaled/clamped delta, got ' + step);
})();

console.log('all assertions passed');

File: fushi/test/lookup/browser_extension_page_scroll_bug732_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

/// BUG-732: Hibiki 的浏览器扩展把 popup.js 作为 content script 注入到所有网页
/// （`<all_urls>`）。popup.js 的 document 'wheel' 监听为「查词弹窗」重写了滚动
/// （按 POPUP_WHEEL_PIXEL_FACTOR=0.24 缩放 + preventDefault）。当滚轮落在宿主页
/// （不在弹窗 shadow 内）时，旧代码回落到 `window.scrollBy`，等于把每个普通网页的
/// 滚动都接管并降速到 24%。
///
/// 根因修复：扩展 content-script 上下文（`chrome.runtime.id` 存在）里，滚轮不在
/// 弹窗 shadow 内时必须回落原生滚动（不 preventDefault）。in-app 弹窗
/// （`flutter_inappwebview`，无 chrome）与「滚轮就在弹窗内」两条路径保持原有缩放滚动。
///
/// 两层守护：
/// ① 行为级——用 Node 真执行 popup.js 的 wheel 监听，断言四种表面下 preventDefault /
///    window.scrollBy / host.scrollBy 的调用。无 node 时 skip。
/// ② 源码级 + 三镜像一致——保证守卫存在、按 `chrome.runtime.id` 判定，且
///    assets/popup 与两份扩展 vendor 镜像逐字节一致（TODO-1267）。
void main() {
  test(
    'BUG-732: extension host-page wheel falls through to native (executes popup.js via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest = File(
        'test/lookup/browser_extension_page_scroll_bug732_test.js',
      );
      expect(jsTest.existsSync(), isTrue,
          reason: 'behavior harness ${jsTest.path} must exist');

      final ProcessResult result = await Process.run(
        nodeExe,
        <String>[jsTest.path],
        workingDirectory: Directory.current.path,
      );

      expect(
        result.exitCode,
        0,
        reason: 'BUG-732 wheel behavior test failed.\n'
            'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
      );
      expect(result.stdout.toString(), contains('all assertions passed'),
          reason: 'behavior harness must reach its success marker');
    },
  );

  test('popup.js wheel guard bails on the extension host page (source guard)',
      () {
    const List<String> popupCopies = <String>[
      'assets/popup/popup.js',
      'assets/browser_extension/vendor/popup.js',
      '../tools/browser-extension/vendor/popup.js',
    ];

    for (final String path in popupCopies) {
      final String src = File(path).readAsStringSync();

      // The extension content-script discriminator: chrome.runtime.id present.
      expect(src.contains('chrome.runtime && chrome.runtime.id'), isTrue,
          reason: '[$path] 必须用 chrome.runtime.id 判定扩展 content-script 上下文');
      // The bail: not over the popup shadow (scroller null) AND in the extension
      // -> return before preventDefault so native scrolling wins.
      expect(src.contains('if (!scroller && inExtensionContentScript) return;'),
          isTrue,
          reason: '[$path] 扩展宿主页（滚轮不在弹窗内）必须回落原生滚动，禁止 preventDefault');
      // The guard must sit BEFORE preventDefault, otherwise the scaled scroll
      // still hijacks the host page.
      final int guardIdx =
          src.indexOf('if (!scroller && inExtensionContentScript) return;');
      final int preventIdx = src.indexOf('e.preventDefault();', guardIdx - 400);
      expect(guardIdx >= 0 && preventIdx > guardIdx, isTrue,
          reason: '[$path] 放行守卫必须在 e.preventDefault() 之前');
    }
  });

  test('the three popup.js copies stay byte-identical (TODO-1267 parity)', () {
    final String a = File('assets/popup/popup.js').readAsStringSync();
    final String b =
        File('assets/browser_extension/vendor/popup.js').readAsStringSync();
    final String c =
        File('../tools/browser-extension/vendor/popup.js').readAsStringSync();
    expect(a == b, isTrue, reason: 'assets/popup 与扩展 vendor 镜像必须逐字节一致');
    expect(a == c, isTrue, reason: 'assets/popup 与 tools 扩展镜像必须逐字节一致');
  });
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}
